### PR TITLE
Remove Headers in favor of Metadata

### LINKED ISSUE #649: Replace Metadata.Trailers and Metadata.Headers with just Metadata
In Headers, path can be removed (just use a different MethodDescriptor; if we really need it it the future, it can be added to CallOptions) and authority can move to CallOptions. At that point, there isn't a distinction between headers and trailers and we can just have "Metadata".

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
Fixes #649

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/HttpUtil.java [hunk 1]
```diff
@@ -44,6 +44,18 @@
...
   /**
+   * {@link io.grpc.Metadata.Key} for the :authority pseudo header.
+   */
+  public static final Metadata.Key<String> AUTHORITY_KEY =
+          Metadata.Key.of(":authority", Metadata.ASCII_STRING_MARSHALLER);
+
+  /**
+   * {@link io.grpc.Metadata.Key} for the :path pseudo header.
+   */
+  public static final Metadata.Key<String> PATH_KEY =
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/HttpUtil.java]: Just delete, we have no need for path, at least not yet.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/HttpUtil.java]: @ejona This is a behavior change and thus an API breakage.  (Client provided headers will override the method name).  Are you sure you want this?

File: core/src/main/java/io/grpc/Metadata.java [hunk 2]
```diff
@@ -297,84 +297,17 @@ private String toStringInternal() {
 
   /**
    * Concrete instance for metadata attached to the start of a call.
+   *
+   * @deprecated use Metadata instead.
    */
+  @Deprecated
```
Comment (nmittler) [inline core/src/main/java/io/grpc/Metadata.java]: *sigh* :/

File: core/src/main/java/io/grpc/internal/HttpUtil.java (hunk 1 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/internal/HttpUtil.java]: +1 ... delete it! quick! :)

**Comment by nmittler** [comment]
LGTM

File: core/src/main/java/io/grpc/internal/HttpUtil.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/HttpUtil.java]: @carl-mastrangelo, are you thinking we should keep it, or preferring to delete it?

File: core/src/main/java/io/grpc/internal/HttpUtil.java [hunk 3]
```diff
@@ -44,6 +44,18 @@
 public final class HttpUtil {
 
   /**
+   * {@link io.grpc.Metadata.Key} for the :authority pseudo header.
+   */
+  public static final Metadata.Key<String> AUTHORITY_KEY =
+          Metadata.Key.of(":authority", Metadata.ASCII_STRING_MARSHALLER);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/HttpUtil.java]: Hmm... so colon in the name should probably be prevented. I guess this shows we need an issue similar to #725 but for metadata keys. Created #824

File: core/src/main/java/io/grpc/internal/HttpUtil.java [hunk 4]
```diff
@@ -44,6 +44,18 @@
 public final class HttpUtil {
 
   /**
+   * {@link io.grpc.Metadata.Key} for the :authority pseudo header.
+   */
+  public static final Metadata.Key<String> AUTHORITY_KEY =
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/HttpUtil.java]: Instead of having this authority key, could we move overriding "authority" to CallOptions?

There is still a question on how to expose the authority on server-side, but that is an orthogonal issue. All users of authority today I think are using it only on client-side.

File: core/src/main/java/io/grpc/Metadata.java (hunk 2 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/Metadata.java]: Yeah I know.  I'd like to break it off too.

File: core/src/main/java/io/grpc/internal/HttpUtil.java (hunk 4 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/HttpUtil.java]: @ejona86  Moved out to call options.

**Comment by carl-mastrangelo** [comment]
Friendly ping

**Comment by ejona86** [comment]
The rebase means I don't know what changed. I'll have to do a full new review.

File: core/src/main/java/io/grpc/inprocess/InProcessTransport.java [hunk 5]
```diff
@@ -106,7 +106,8 @@ public void run() {
 
   @Override
   public synchronized ClientStream newStream(MethodDescriptor<?, ?> method,
-      Metadata.Headers headers, ClientStreamListener clientStreamListener) {
+      Metadata headers, ClientStreamListener clientStreamListener) {
+    // FIXME: Remove authority
```
Comment (ejona86) [inline core/src/main/java/io/grpc/inprocess/InProcessTransport.java]: Remove authority here?

File: netty/src/main/java/io/grpc/netty/NettyClientTransport.java [hunk 6]
```diff
@@ -119,11 +119,12 @@ public void ping(PingCallback callback, Executor executor) {
   }
 
   @Override
-  public ClientStream newStream(MethodDescriptor<?, ?> method, Metadata.Headers headers,
+  public ClientStream newStream(MethodDescriptor<?, ?> method, Metadata headers,
       ClientStreamListener listener) {
     Preconditions.checkNotNull(method, "method");
     Preconditions.checkNotNull(headers, "headers");
     Preconditions.checkNotNull(listener, "listener");
+ // FIXME: Remove authority
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyClientTransport.java]: Remove the authority key here, and if it was non-null, pass it into `convertClientHeaders()` instead of the class-field `authority`.

File: okhttp/src/main/java/io/grpc/okhttp/Headers.java [hunk 7]
```diff
@@ -74,9 +75,11 @@
     // Set GRPC-specific headers.
     okhttpHeaders.add(SCHEME_HEADER);
     okhttpHeaders.add(METHOD_HEADER);
-    String authority = headers.getAuthority() != null ? headers.getAuthority() : defaultAuthority;
+
+    String authority = headers.containsKey(AUTHORITY_KEY)
```
Comment (ejona86) [inline okhttp/src/main/java/io/grpc/okhttp/Headers.java]: Remove the authority as well.

File: netty/src/main/java/io/grpc/netty/NettyClientTransport.java (hunk 6 above)
Comment (carl-mastrangelo) [inline netty/src/main/java/io/grpc/netty/NettyClientTransport.java]: Done.  It seems like Path, Scheme, and method should also be removed, but meh.

File: okhttp/src/main/java/io/grpc/okhttp/Headers.java (hunk 7 above)
Comment (carl-mastrangelo) [inline okhttp/src/main/java/io/grpc/okhttp/Headers.java]: Done

File: core/src/main/java/io/grpc/inprocess/InProcessTransport.java (hunk 5 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/inprocess/InProcessTransport.java]: Done.

**Comment by carl-mastrangelo** [comment]
@ejona86  PTAL

**Comment by ejona86** [comment]
@carl-mastrangelo LGTM
